### PR TITLE
Added MNE-BIDS utils to handle BIDS compliant datasets

### PR DESCRIPTION
**Pull request by estefanysuarez**
### Add MNE and BIDS utilities for EEG and iEEG data management
This PR introduces a new `bids_utils` module with functions to work with Brain Imaging Data Structure (BIDS) compliant neuroimaging data, particularly for EEG and iEEG recordings, and also adds some functions to the `mne_utils` module. 

Key Features from `bids_utils` module:
- Discover EEG and iEEG recordings from BIDS directories or a list of candidate files (e.g., a list of files fetched from a S3 bucket)
- Group recordings based on fixed BIDS entities
- Check if data files matching a recording exist locally
- Build MNE-BIDS paths for seamless integration with the MNE ecosystem
- Load files specifically associated to BIDS datasets (e.g., `participants.tsv` or `.json` sidecars)
- Support for multiple BIDS-compliant formats (.edf, .vhdr, .set, .bdf, .nwb)
- Testing: Comprehensive test coverage for all functions including edge cases and error conditions.

New features in `mne_utils` module:
- A function that concatenates multiple iEEG or EEG recordings
- Extended functionality for `extract_channels`
- Tests were updated to include the new features

### Details of extended functionality for `extract_channels` in mne_utils
- The new function works for both EEG and iEEG MNE Raw objects. If `channels_name_mapping` or `channels_type_mapping` dictioneries are passed, then channel names and types will be remapped. This function can now replace `_build_channels` in `OpenNeuroPipeline` class. 
- A new 'bad' flag is added to identify bad channels
- Coordinates will be extracted if the exist in the MNE Raw object or if they are passed as a dictionary mapping channel names to x, y, z coordinates

**Note:** Since Open Neuro datasets are BIDS compliant and MNE compatible, these utilities can be reused for the OpenNeuro Pipelines ✨

## Summary by CodeRabbit

* **New Features**
  * Discover and validate BIDS-formatted EEG and iEEG datasets, group recordings by BIDS entities, and check recording file existence.
  * Load JSON sidecars and participants metadata; retrieve subject age/sex.
  * New recording-concatenation API with measurement-date handling and offset checks; revised full-signal and channel extraction with configurable name/type/position mappings and bad-channel support.

* **Tests**
  * Extensive test coverage for BIDS utilities and the updated signal extraction/concatenation behaviors.

### CODE REVIEW

File: brainsets/utils/bids_utils.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+    recordings: list[dict],
+    fixed_entities: Optional[list[str]] = None,
+) -> dict[str, list[dict]]:
+    """Group BIDS-compliant recordings by specified fixed entities.
+
+    Group keys are constructed using only the entities listed in
+    `fixed_entities`; all other entities are implicitly allowed to vary within
+    a group.
+
+    By default (`fixed_entities=None`), groups are created by all entities except 'run'.
```
[the remark this answers is not in the document]
Comment (ghazalehran) [reply brainsets/utils/bids_utils.py]: if I understand correctlly fixed_entities=None currently means “group by all entities except run”, which is convenient, but not necessarily the least surprising behavior. I would be fine either making that default more explicit in the API/docs or revisiting the default semantics in a follow-up.

File: brainsets/utils/mne_utils.py [hunk 2]
```diff
@@ -49,73 +54,198 @@ def extract_measurement_date(
...
+    elif on_offset == "warn":
+        # A more efficient way: use zip to iterate through consecutive pairs directly.
+        for (idx1, _, date1), (idx2, _, date2) in zip(
+            sorted_recordings, sorted_recordings[1:]
+        ):
+            offset = (date2 - date1).total_seconds()
+            if offset > 3600:
+                warnings.warn(
+                    f"Offset between recordings {idx1} and {idx2} is greater than 1 hour: {offset} seconds"
+                )
```
[the remark this answers is not in the document]
Comment (estefanysuarez) [reply brainsets/utils/mne_utils.py]: The code now handles both missing and time-aware meas_dates. Done ✅

File: brainsets/utils/mne_utils.py [hunk 3]
```diff
@@ -49,73 +56,228 @@ def extract_measurement_date(
...
-def extract_eeg_signal(
-    recording_data: "mne.io.Raw",
-) -> RegularTimeSeries:
-    """Extract the EEG signal as a RegularTimeSeries from MNE Raw data.
+def concatenate_recordings(
+    recordings: list["mne.io.Raw"],
+    max_offset: float = 1.0,
+    on_mismatch: Literal["ignore", "warn", "raise"] = "raise",
+    on_offset: Literal["ignore", "warn", "raise"] = "warn",
+    on_missing_meas_date: Literal["ignore", "warn", "raise"] = "warn",
```
[the remark this answers is not in the document]
Comment (estefanysuarez) [reply brainsets/utils/mne_utils.py]: DONE ✅

**Review by oumgharbi — COMMENTED** [review]
Thank you @estefanysuarez for these utils. 
My only concern is that the `bids_utils` assumes BIDS compliance (which makes perfect sense), but maybe adding a BIDS validator like [this one](https://github.com/bids-standard/bids-validator) may help avoid certain errors in the future, on certain published datasets.

**Review by ghazalehran — APPROVED** [review]

File: brainsets/utils/bids_utils.py [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+
+For more information about BIDS, see the BIDS specification: https://bids-specification.readthedocs.io/en/stable/
+"""
+
+from collections import defaultdict
+from typing import Optional
+from enum import Enum
+import warnings
+from pathlib import Path
+import warnings
```
Comment (milosobral) [inline brainsets/utils/bids_utils.py]: Unused import

File: brainsets/utils/bids_utils.py [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+
+    This searches for any file belonging to the recording in the proper BIDS directory structure,
+    matching any of the supported file extensions. It supports all BIDS-compliant formats (e.g., .edf, .vhdr, .set, .bdf, .eeg, .nwb)
+    plus .fif for MNE-processed files.
+
+    Args:
+        bids_root: BIDS root directory (e.g., '/path/to/bids/root')
+        recording_id: Recording identifier (e.g., 'sub-1_task-Sleep_acq-headband')
+        modality: Modality (supported values: 'eeg', 'ieeg')
+        extensions: Set of allowed file extensions (e.g., EEG_EXTENSIONS or IEEG_EXTENSIONS)
```
Comment (vinamarora8) [inline brainsets/utils/bids_utils.py]: Needs to be updated to match the fn signature

File: brainsets/utils/mne_utils.py [hunk 6]
```diff
@@ -49,126 +55,568 @@ def extract_measurement_date(
...
+    }
+
+    return channel_names_mapping
+
+
+def _validate_channel_types_mapping(
+    raw_data: "mne.io.BaseRaw",
+    channel_names_mapping: dict[str, str] | None = None,
+    channel_types_mapping: dict[str, list[str]] | None = None,
+) -> dict[str, str]:
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply brainsets/utils/mne_utils.py]: (maybe I'm wrong here!) I think this fn isn't performing a validation, but "transposing" the input `channel_types_mapping` from {type: [names]} to {name: type}. If this is the case, then it should be renamed to something like `_transpose_channel_types_mapping`?

File: brainsets/utils/bids_utils.py (hunk 5 above)
Comment (estefanysuarez) [reply brainsets/utils/bids_utils.py]: DONE ✅

File: brainsets/utils/mne_utils.py [hunk 7]
```diff
@@ -49,126 +55,568 @@ def extract_measurement_date(
...
+def _validate_channel_pos_mapping(
+    raw_data: "mne.io.BaseRaw",
+    channel_names_mapping: dict[str, str] | None = None,
+    channel_pos_mapping: dict[str, np.ndarray] | None = None,
+) -> dict[str, np.ndarray] | None:
+    """Validate and return a channel position mapping.
+
+    Returns a mapping of original channel names to their positions (x, y, z).
+    If no position mapping is provided, returns an empty dict. If a position
+    mapping is provided, it resolves whether the mapping keys refer to
```
[the remark this answers is not in the document]
Comment (estefanysuarez) [reply brainsets/utils/mne_utils.py]: Updated docstrings to reflect the right behavior. DONE ✅

File: brainsets/utils/mne_utils.py [hunk 8]
```diff
@@ -49,126 +55,568 @@ def extract_measurement_date(
...
+            end=np.array([(len(signal) - 1) / sfreq]),
         ),
     )
 
 
 def extract_channels(
-    recording_data: "mne.io.Raw",
+    recording_data: "mne.io.BaseRaw",
+    channel_names_mapping: dict[str, str] | None = None,
+    channel_types_mapping: dict[str, list[str]] | None = None,
```
[the remark this answers is not in the document]
Comment (estefanysuarez) [reply brainsets/utils/mne_utils.py]: KempSleepEDF was updated too!

File: brainsets/utils/mne_utils.py [hunk 9]
```diff
@@ -533,7 +537,7 @@ def _validate_channel_names_mapping(
 
 def _validate_channel_types_mapping(
     raw_data: "mne.io.BaseRaw",
-    channel_names_mapping: dict[str, str] | None = None,
+    channel_names_mapping: dict[str, str],
     channel_types_mapping: dict[str, list[str]] | None = None,
```
Comment (vinamarora8) [inline brainsets/utils/mne_utils.py]: I think the correct type for `channel_types_mapping` now is `dict[str, str]`

**Review by vinamarora8 — COMMENTED** [review]
Thanks for the quick changes!! Small comments on the updates. Looks good otherwise!

File: brainsets/utils/bids_utils.py [hunk 10]
```diff
@@ -657,20 +653,25 @@ def _is_bids_root(path: str | Path) -> bool:
...
+    Raises:
+        ValueError: If modality is not one of the supported values in SUPPORTED_MODALITIES.
+
+    Examples:
+        >>> _validate_modality("eeg")  # No error
+        >>> _validate_modality("ieeg")  # No error
+        >>> _validate_modality("unsupported_modality")  # Raises ValueError
+    """
+    if modality.lower() not in SUPPORTED_MODALITIES:
+        raise ValueError(
```
[the remark this answers is not in the document]
Comment (estefanysuarez) [reply brainsets/utils/bids_utils.py]: yeeeah, you are right! Also updated the respective test. DONE ✅

**Comment by estefanysuarez** [comment]
> Thanks for making the changes! PR is good to go! @milosobral might also have to approve to clear the "Changes Requested"

Thank you @vinamarora8 !
